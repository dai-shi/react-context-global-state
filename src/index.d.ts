import * as React from 'react';

export type StateProviderProps = {
  children?: React.ReactNode;
};

export type Update<T> = ((v: T) => T) | T;

export type StateConsumerProps<S, N extends keyof S> = {
  name: N;
  children?: (value: S[N], updater: (u: Update<S[N]>) => void) => React.ReactNode;
};

export type StateConsumerType<S, N extends keyof S> = React.ComponentType<StateConsumerProps<S, N>>;

export type SetGlobalState<S> = <N extends keyof S>(
  name: N,
  update: Update<S[N]>,
) => void;

export type GetGlobalState<S> = <N extends keyof S>(
  name: N,
) => S[N];

export const createGlobalState: <S extends {}, N extends keyof S>(initialState: S) => {
  StateProvider: React.ComponentType<StateProviderProps>;
  StateConsumer: StateConsumerType<S, N>;
  setGlobalState: SetGlobalState<S>;
  getGlobalState: GetGlobalState<S>;
};
