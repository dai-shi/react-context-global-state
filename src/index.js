import React, { createElement as h } from 'react';

const isFunction = (fn) => (typeof fn === 'function');

export const createGlobalState = (initialState) => {
  const stateItemConsumers = {};
  const stateItemUpdateListeners = {};
  const stateItemUpdaters = {};
  const stateItemValues = initialState;
  let StateProvider = ({ children }) => h(React.Fragment, null, children);
  Object.keys(initialState).forEach((name) => {
    const { Provider, Consumer } = React.createContext({
      value: initialState[name],
      update: () => { throw new Error('cannot update initial value'); },
    });
    stateItemConsumers[name] = ({ children }) => h(Consumer, null,
      ({ value, update }) => children(value, update));
    stateItemUpdateListeners[name] = [];
    stateItemUpdaters[name] = (func) => {
      stateItemUpdateListeners[name].forEach((listener) => listener(func));
    };
    const InnerProvider = StateProvider;
    StateProvider = class extends React.PureComponent {
      constructor() {
        super();
        stateItemUpdateListeners[name].push((funcOrValue) => {
          let newValue;
          if (isFunction(funcOrValue)) {
            const { value: oldValue } = this.state;
            newValue = funcOrValue(oldValue);
          } else {
            newValue = funcOrValue;
          }
          this.setState({ value: newValue });
          stateItemValues[name] = newValue;
        });
        this.state = { value: initialState[name], update: stateItemUpdaters[name] };
      }

      render() {
        const { children } = this.props;
        return h(Provider, { value: this.state },
          h(InnerProvider, null, children));
      }
    };
  });
  const StateConsumer = ({ name, children }) => stateItemConsumers[name]({ children });
  return {
    StateProvider,
    StateConsumer,
    setGlobalState: (name, update) => stateItemUpdaters[name](update),
    getGlobalState: (name) => stateItemValues[name],
  };
};
