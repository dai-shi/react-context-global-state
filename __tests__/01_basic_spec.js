/* eslint-env jest */

import React from 'react';
import { configure, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';

import { createGlobalState } from '../src/index';

configure({ adapter: new Adapter() });

describe('basic spec', () => {
  it('should have a function', () => {
    expect(createGlobalState).toBeDefined();
  });

  it('should create a component with a global state', () => {
    const initialState = {
      count1: 0,
    };
    const { StateProvider, StateConsumer } = createGlobalState(initialState);
    const Counter = () => (
      <StateConsumer name="count1">
        {(value, update) => (
          <div>
            <span>{value}</span>
            <button type="button" onClick={() => update(value + 1)}>+1</button>
          </div>
        )}
      </StateConsumer>
    );
    const App = () => (
      <StateProvider>
        <div className="first">
          <Counter />
        </div>
        <div className="second">
          <Counter />
        </div>
      </StateProvider>
    );
    const wrapper = mount(<App />);
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.find('.first button').simulate('click');
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
