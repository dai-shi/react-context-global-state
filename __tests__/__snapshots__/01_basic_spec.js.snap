// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic spec should create a component with a global state 1`] = `
<App>
  <StateProvider>
    <StateProvider>
      <div
        className="first"
      >
        <Counter>
          <StateConsumer
            name="count1"
          >
            <div>
              <span>
                0
              </span>
              <button
                onClick={[Function]}
                type="button"
              >
                +1
              </button>
            </div>
          </StateConsumer>
        </Counter>
      </div>
      <div
        className="second"
      >
        <Counter>
          <StateConsumer
            name="count1"
          >
            <div>
              <span>
                0
              </span>
              <button
                onClick={[Function]}
                type="button"
              >
                +1
              </button>
            </div>
          </StateConsumer>
        </Counter>
      </div>
    </StateProvider>
  </StateProvider>
</App>
`;

exports[`basic spec should create a component with a global state 2`] = `
<App>
  <StateProvider>
    <StateProvider>
      <div
        className="first"
      >
        <Counter>
          <StateConsumer
            name="count1"
          >
            <div>
              <span>
                1
              </span>
              <button
                onClick={[Function]}
                type="button"
              >
                +1
              </button>
            </div>
          </StateConsumer>
        </Counter>
      </div>
      <div
        className="second"
      >
        <Counter>
          <StateConsumer
            name="count1"
          >
            <div>
              <span>
                1
              </span>
              <button
                onClick={[Function]}
                type="button"
              >
                +1
              </button>
            </div>
          </StateConsumer>
        </Counter>
      </div>
    </StateProvider>
  </StateProvider>
</App>
`;
